docs/website/blog/2022-06-30-binding-database-service-without-sbo.md

```diff
@@ -0,0 +1,170 @@
+
+## Setting up the namespace
+2. We will create a new namespace to deploy our application in, with the help of odo.
+```sh
+odo create namespace restapi-mongodb
+```
+
+
+## Setting up the mongodb microservice
+We are going to use the Percona's operator for creating our mongodb database. For the sake of simplicity, we will use Helm to deploy our mongodb operator and the service.
+
+3. Add the Percona’s Helm charts repository and make your Helm client up to date with it:
+```sh
+helm repo add percona https://percona.github.io/percona-helm-charts/ && helm repo update
+```
+
+4. Install Percona Operator for MongoDB:
+```sh
+helm install my-op percona/psmdb-operator
+```
+Wait for the pod to come up:
+```sh
+$ kubectl get pods
+NAME                                    READY   STATUS    RESTARTS   AGE
+my-op-psmdb-operator-69d88f479c-7hj5m   1/1     Running   0          3m34s
+```
+
+5. Install Percona server for MongoDB from our local `psmdb-db` Helm chart.
+```sh
+helm install my-db ./psmdb-db
+```
+
+```sh
+$ helm install my-db ./psmdb-db
+
+NAME: my-db
+LAST DEPLOYED: Mon Jul  4 13:49:26 2022
+NAMESPACE: restapi-mongodb
+STATUS: deployed
+REVISION: 1
+TEST SUITE: None
+NOTES:
+To get a MongoDB prompt inside your new cluster you can run:
+
+  ADMIN_USER=$(kubectl -n restapi-mongodb get secrets minimal-cluster -o jsonpath="{.data.MONGODB_USER_ADMIN_USER}" | base64 --decode)
+  ADMIN_PASSWORD=$(kubectl -n restapi-mongodb get secrets minimal-cluster -o jsonpath="{.data.MONGODB_USER_ADMIN_PASSWORD}" | base64 --decode)
+
+And then for replica set:
+  $ kubectl run -i --rm --tty percona-client --image=percona/percona-server-mongodb:4.4 --restart=Never -- mongo "mongodb+srv://${ADMIN_USER}:${ADMIN_PASSWORD}@minimal-cluster-rs0.restapi-mongodb.svc.cluster.local/admin?replicaSet=rs0&ssl=false"
+
+Or for sharding setup:
+  $ kubectl run -i --rm --tty percona-client --image=percona/percona-server-mongodb:4.4 --restart=Never -- mongo "mongodb://${ADMIN_USER}:${ADMIN_PASSWORD}@minimal-cluster-mongos.restapi-mongodb.svc.cluster.local/admin?ssl=false"
+```
+
+Wait for the pods to come up, this might take a few minutes:
+```sh
+$ kubectl get pods
+NAME                                    READY   STATUS    RESTARTS   AGE
+minimal-cluster-cfg-0                   1/1     Running   0          3m6s
+minimal-cluster-mongos-0                1/1     Running   0          3m5s
+minimal-cluster-rs0-0                   1/1     Running   0          3m5s
+my-op-psmdb-operator-69d88f479c-7hj5m   1/1     Running   0          3m34s
+```
+
+## Download the devfile.yaml
+6. Run `odo init` to fetch the necessary devfile.
+```sh
+odo init --devfile go --name places
+```
+
+7. Edit the 'runtime' component container in devfile to add information such as username, password, and host required to connect to the mongodb service.
+```yaml
+components:
+- container:
+    ...
+    ...
+    env:
+    - name: username
+      value: userAdmin
+    - name: password
+      value: userAdmin123456
```

Comment:
Does devfile support dynamically obtaining values? I mean if we can put `kubectl get secrets minimal-secret -ojsonpath='{.data.username}' | base64 -d` inside the devfile, we can avoid having to hard code password.

Reply:
I could add a note acknowledging that it's a bad practice, but this is for example purpose only, or I could find a way to fetch password from the code itself to avoid this practice altogether.

Reply:
Yes, you could start with this solution, by indicating from the very beginning this is not a secure solution, and then continue with more secure solutions (and probaly more difficult ones). That would show the interest of using SBO.

- The simplest solution would be to have an `envFrom` field in the Devfile's container, but this field does not exist. Is it worth asking Devfile team to add it? I'm not sure
- You could pass the secret name and the secret keys as environment variables, and the program would be responsible for getting the values from the secret. That means the program needs access to the Kubernetes API, which opens security breaches

Reply:
>The simplest solution would be to have an envFrom field in the Devfile's container, but this field does not exist. Is it worth asking Devfile team to add it? I'm not sure

When you say `envFrom`, do you mean devfile can evaluate the value from the environment variable exposed in your local system?
```yaml
env:
- name: password
  value:
    envFrom: $MONGODB_PASSWORD
```
Something like this? Or more like this https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/pods/inject/pod-secret-envFrom.yaml?


>You could pass the secret name and the secret keys as environment variables, and the program would be responsible for getting the values from the secret. That means the program needs access to the Kubernetes API, which opens security breaches

I think asking the program to be responsible for fetching everything would defeat the purpose of this blog. I think I'll go with a note for now, and it's like you said, it might encourage users to try SBO.

Reply:
```
> 
> Something like this? Or more like this https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/pods/inject/pod-secret-envFrom.yaml?
> 

Yes, like this one

Reply:
But then devfile would have to interact with the k8s API and that does not make a lot of sense to me from the design standpoint. I don't think devfile is dependent on the K8s API right now, and I'm not sure if they'd be willing to create a dependency now. Can we take this up in cabal?